wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,542 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Check and set KeyId
+	switch {
+	case os.Getenv(EnvPkcs11WrapperKeyId) != "" && !opts.Options.WithDisallowEnvVars:
+		k.keyId = os.Getenv(EnvPkcs11WrapperKeyId)
+	case os.Getenv(EnvVaultPkcs11SealKeyId) != "" && !opts.Options.WithDisallowEnvVars:
+		k.keyId = os.Getenv(EnvVaultPkcs11SealKeyId)
+	case opts.WithKeyId != "":
+		k.keyId = opts.WithKeyId
+	default:
+		return nil, fmt.Errorf("key id not found (env or config) for pkcs11 wrapper configuration")
+	}
+
+	// Set and check k.client
+	if k.client == nil {
+		k.client = &pkcs11KMS{}
+
+		if !opts.Options.WithDisallowEnvVars && os.Getenv("PKCS11_SLOT") != "" {
+			var err error
+			var slot uint64
+			slot, err = strconv.ParseUint(os.Getenv("PKCS11_SLOT"), 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			opts.withSlot = uint(slot)
+		}
+		if k.client.slot == 0 {
+			k.client.slot = opts.withSlot
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.pin = os.Getenv("PKCS11_PIN")
+		}
+		if k.client.pin == "" {
+			k.client.pin = opts.withPin
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.module = os.Getenv("PKCS11_MODULE")
+		}
+		if k.client.module == "" {
+			k.client.module = opts.withModule
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.label = os.Getenv("PKCS11_LABEL")
+		}
+		if k.client.label == "" {
+			k.client.label = opts.withLabel
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			mechanismName := os.Getenv("PKCS11_MECHANISM")
+			if mechanismName != "" {
+				k.client.mechanism, err = MechanisFromString(mechanismName)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		if k.client.mechanism == 0 {
+			if opts.withMechanism != "" {
+				k.client.mechanism, err = MechanisFromString(opts.withMechanism)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		k.client.keyId = k.keyId
+
+		p := pkcs11.New(k.client.module)
+		err := p.Initialize()
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+		}
+		defer p.Destroy()
+		defer p.Finalize()
+
+		session, err := p.OpenSession(k.client.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open session: %w", err)
+		}
+		defer p.CloseSession(session)
+
+		err = p.Login(session, pkcs11.CKU_USER, k.client.pin)
+		if err != nil {
+			return nil, fmt.Errorf("failed to login: %w", err)
+		}
+		defer p.Logout(session)
+
+	}
+	// Store the current key id. If using a key alias, this will point to the actual
+	// unique key that that was used for this encrypt operation.
+	k.currentKeyId.Store(k.keyId)
+
+	// Map that holds non-sensitive configuration info
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["kms_key_id"] = k.keyId
+	wrapConfig.Metadata["slot"] = strconv.Itoa(int(k.client.slot))
+	if k.client.label != "" {
+		wrapConfig.Metadata["label"] = k.client.label
```

Comment:
Per the docs, I think this [actually backwards](https://developer.hashicorp.com/vault/docs/configuration/seal/pkcs11#key_label) -- I think `label` is required, `key_id` is optional.